File: application/app/src/main/java/com/yahyeet/boardbook/presenter/game/gamedetail/GameDetailAdapter.java
package com.yahyeet.boardbook.presenter.game.gamedetail;


import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.yahyeet.boardbook.R;
import com.yahyeet.boardbook.model.entity.GameRole;
import com.yahyeet.boardbook.model.entity.GameTeam;

import java.util.ArrayList;
import java.util.List;

/**
 * Adapter for the game detail recycler view
 */
public class GameDetailAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	private static final int FIRST_LIST_ITEM_VIEW = 1;
	private static final int SECOND_LIST_ITEM_VIEW = 2;

	private List<GameTeam> firstList;
	private List<List<GameRole>> secondList;

	private List<String> allNames = new ArrayList<>();

	public GameDetailAdapter(List<GameTeam> firstList, List<List<GameRole>> secondList) {
		this.firstList = firstList;
		this.secondList = secondList;

		for (GameTeam team : firstList){
			allNames.add(team.getName());
			if(!team.getRoles().isEmpty()){
				for(GameRole role : team.getRoles())
					allNames.add(role.getName());
			}
		}

	}

	public class ViewHolder extends RecyclerView.ViewHolder {
		// List items of first list
		private TextView tvTeamName;

		// List items of second list
		private TextView tvRoleName;


		public ViewHolder(final View itemView) {
			super(itemView);

			// Get the view of the elements of first list
			tvTeamName = itemView.findViewById(R.id.gameDetailTeamName);

			// Get the view of the elements of second list
			tvRoleName =  itemView.findViewById(R.id.gameDetailRoleName);
		}

		void bindViewSecondList(int pos) {

			final String description = allNames.get(pos);

			tvRoleName.setText(description);
		}

		void bindViewFirstList(int pos) {

			final String description = allNames.get(pos);

			tvTeamName.setText(description);
		}


	}

	private class FirstListItemViewHolder extends ViewHolder {
		public FirstListItemViewHolder(View itemView) {
			super(itemView);
		}
	}


	private class SecondListItemViewHolder extends ViewHolder {
		public SecondListItemViewHolder(View itemView) {
			super(itemView);
		}
	}

	@NonNull
	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

		View v;

		if (viewType == FIRST_LIST_ITEM_VIEW) {
			v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_team, parent, false);
			FirstListItemViewHolder vh = new FirstListItemViewHolder(v);
			return vh;

		} else {
			// SECOND_LIST_ITEM_VIEW
			v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_role, parent, false);
			SecondListItemViewHolder vh = new SecondListItemViewHolder(v);
			return vh;
		}
	}

	@Override
	public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {

		try {
			if (holder instanceof SecondListItemViewHolder) {
				SecondListItemViewHolder vh = (SecondListItemViewHolder) holder;
				vh.bindViewSecondList(position);

			} else if (holder instanceof FirstListItemViewHolder) {
				FirstListItemViewHolder vh = (FirstListItemViewHolder) holder;
				vh.bindViewFirstList(position);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	@Override
	public int getItemCount() {

		int firstListSize = 0;
		int secondListSize = 0;

		if (secondList == null && firstList == null) return 0;


		if (firstList != null)
			firstListSize = firstList.size();


		for (List<GameRole> list : secondList) {
			for (GameRole ignored : list)
				secondListSize++;
		}

		return firstListSize + secondListSize;
	}

	@Override
	public int getItemViewType(int position) {


		if (secondList == null && firstList == null)
			return super.getItemViewType(position);


		if (secondList != null) {
			int i = (int) secondList.stream().filter(list -> !list.isEmpty()).count();
			if (i == 0)
				return FIRST_LIST_ITEM_VIEW;
		}


		if (firstList == null)
			return SECOND_LIST_ITEM_VIEW;


		List<Integer> sizeList = new ArrayList<>();

		for (List<GameRole> list : secondList)
			if(!list.isEmpty())
				sizeList.add(list.size());

		int desiredPosition = position;

		int sumOfPreviousTeamSizes = 0;
		for (int i = 0; i < sizeList.size(); i++) {
			if (sumOfPreviousTeamSizes + sizeList.get(i) >= position) {
				desiredPosition = position - sumOfPreviousTeamSizes - i;
				break;
			} else {
				sumOfPreviousTeamSizes += sizeList.get(i);
			}
		}

		if (desiredPosition == 0)
			return FIRST_LIST_ITEM_VIEW;
		else
			return SECOND_LIST_ITEM_VIEW;
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/presenter/game/gamedetail/GameDetailAdapter.kt
package com.yahyeet.boardbook.presenter.game.gamedetail


import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import androidx.recyclerview.widget.RecyclerView

import com.yahyeet.boardbook.R
import com.yahyeet.boardbook.model.entity.GameRole
import com.yahyeet.boardbook.model.entity.GameTeam

import java.util.ArrayList

class GameDetailAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private var firstList: List<GameTeam>? = ArrayList()
    private var secondList: List<List<GameRole>>? = ArrayList()

    private val allNames = ArrayList<String>()

    constructor() {}

    constructor(firstList: List<GameTeam>, secondList: List<List<GameRole>>) {
        this.firstList = firstList
        this.secondList = secondList

        for (team in firstList) {
            allNames.add(team.name)
            for (role in team.roles)
                allNames.add(role.name)
        }

    }

    fun setFirstList(firstList: List<GameTeam>) {
        this.firstList = firstList
    }

    fun setSecondList(secondList: List<List<GameRole>>) {
        this.secondList = secondList
    }

    open inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        // List items of first list
        private val tvTeamName: TextView

        // List items of second list
        private val tvRoleName: TextView


        init {

            // Get the view of the elements of first list
            tvTeamName = itemView.findViewById<View>(R.id.gameDetailTeamName) as TextView

            // Get the view of the elements of second list
            tvRoleName = itemView.findViewById<View>(R.id.gameDetailRoleName) as TextView
        }

        fun bindViewSecondList(pos: Int) {

            val description = allNames[pos]

            tvRoleName.text = description
        }

        fun bindViewFirstList(pos: Int) {

            val description = allNames[pos]

            tvTeamName.text = description
        }


    }

    private inner class FirstListItemViewHolder(itemView: View) : ViewHolder(itemView)


    private inner class SecondListItemViewHolder(itemView: View) : ViewHolder(itemView)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        val v: View

        if (viewType == FIRST_LIST_ITEM_VIEW) {
            v = LayoutInflater.from(parent.context).inflate(R.layout.list_item_team, parent, false)
            return FirstListItemViewHolder(v)

        } else {
            // SECOND_LIST_ITEM_VIEW
            v = LayoutInflater.from(parent.context).inflate(R.layout.list_item_role, parent, false)
            return SecondListItemViewHolder(v)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        try {
            if (holder is SecondListItemViewHolder) {
                holder.bindViewSecondList(position)

            } else if (holder is FirstListItemViewHolder) {
                holder.bindViewFirstList(position)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun getItemCount(): Int {

        var firstListSize = 0
        var secondListSize = 0

        if (secondList == null && firstList == null) return 0


        if (firstList != null)
            firstListSize = firstList!!.size


        for (list in secondList!!) {
            for (role in list)
                secondListSize++
        }

        return firstListSize + secondListSize
    }

    override fun getItemViewType(position: Int): Int {


        if (secondList == null && firstList == null)
            return super.getItemViewType(position)

        if (secondList == null)
            return FIRST_LIST_ITEM_VIEW

        if (firstList == null)
            return SECOND_LIST_ITEM_VIEW


        val sizeList = ArrayList<Int>()

        for (list in secondList!!)
            sizeList.add(list.size)

        var desiredPosition = position

        var sumOfPreviousTeamSizes = 0
        for (i in sizeList.indices) {
            if (sumOfPreviousTeamSizes + sizeList[i] >= position) {
                desiredPosition = position - sumOfPreviousTeamSizes - i
                break
            } else {
                sumOfPreviousTeamSizes += sizeList[i]
            }
        }

        return if (desiredPosition == 0)
            FIRST_LIST_ITEM_VIEW
        else
            SECOND_LIST_ITEM_VIEW
    }

    companion object {

        private val FIRST_LIST_ITEM_VIEW = 1
        private val SECOND_LIST_ITEM_VIEW = 2
    }
}
